### ISSUE TITLE
NPM cannot find package.json in non-root directory

### ISSUE THREAD

**Issue by davidmather** [Microsoft/WSL#2758]
* Your Windows build number:
  * Microsoft Windows [Version 10.0.16299.98]

* What you're doing and what's happening:  

  * I followed https://github.com/Microsoft/WSL/issues/1512 to install npm which worked properly.

  * I am trying to run npm install while specifying a directory path for the package.json file. I want to put Linux only dependencies into a package.json in another directory so when I run npm install from command prompt npm won't try to install the windows versions of those dependencies. 
    ```
     David@MSI:/mnt/c/Users/David/Desktop/we$ npm install ./mnt/c/Users/David/Desktop/we/linux angular --save
     npm ERR! code ENOLOCAL
     npm ERR! Could not install from "mnt/c/Users/David/Desktop/test/linux" as it does not contain a    package.json file.
    ```

* What's wrong / what should be happening instead:
    * npm cannot find the directory

* Strace of the failing command, 
https://gist.github.com/davidmather/84abf3a23d7268b519ee1163fd7510e7

**Comment by therealkenc** [comment]
> David@MSI:/mnt/c/Users/David/Desktop/we$ npm install ./mnt/c/Users/David/Desktop/we/linux

That is path `/mnt/c/Users/David/Desktop/we/mnt/c/Users/David/Desktop/we/linux`, which does not exist and does not contain a `package.json`.

```
cd /mnt/c/Users/David/Desktop/we/linux
npm install angular --save
```
...assuming /mnt/c/Users/David/Desktop/we/linux is a directory containing a `package.json`

**Comment by davidmather** [comment]
Yes /mnt/c/Users/David/Desktop/we/Linux is a path containing a package.json. 

I don't want to cd because that will install in the same directory I run it in. 

I want to use a package.json that is not in the current working directory it is the the equivalent of

``
PS C:\Users\David\Desktop\we> npm install C:\Users\David\Desktop\we\linux angular --save-dev
``

from powershell which installs to the node-modules directory in the \we folder and not in the \linux folder.

**Comment by therealkenc** [comment]
I think you want:
```
mkdir -p /mnt/c/Users/David/Desktop/we/linux/node_modules
npm install --prefix /mnt/c/Users/David/Desktop/we/linux angular --save-dev
```
But a nodejs group would probably give you more informed answers than myself.

**Comment by davidmather** [comment]
No that isn't it. A friend of mine was complaining because he wanted to use Linux specific node_modules but in his project. There should be one node_modules folder for all the modules which is in the project root in this case I called it 'we' because those are the keys I typed quickly trying to figure it out for him.  Now if I have both windows only modules and Linux only modules in the root package.json what happens is that if I run npm install on windows then npm gets the windows versions. Which can break things because you are downloading the windows version of something you need the Linux version of or vice versa. Npm install directoryPath allows you to specify the location of the package.json file to use so I want to separate my files into /Windows and /Linux for those files that are system specific but I want the modules to install to the root project directory in this case 'we'. So I have we/node_modules with all the npm modules and we/Linux contains a single package.json file specifying modules I need to run under Linux. Problem is npm install directoryPath isn't finding the directory for me for some reason it is telling me the folder containing the package.json doesn't exist.

**Comment by therealkenc** [comment]
> Npm install directoryPath allows you to specify the location of the package.json

The directory path you specified was:
```
/mnt/c/Users/David/Desktop/we/mnt/c/Users/David/Desktop/we/linux
```
Try dropping the `./`, because you are asking for a path starting in `$PWD`. But again, you are going to get better answers on `npm` usage in a nodejs forum.

**Comment by davidmather** [comment]
Yeah I tried dropping the ./ it didn't work 

I dunno if I would get better answers on a nodejs forum since it is only not working on wsl

**Comment by therealkenc** [comment]
>  it is only not working on wsl

To clarify: If you have a CIFS mount `/mnt/c` on Real Linux™ it works with the same commands using the same version of `npm`? I'll give it a try tomorrow (Christmas stuff tonight). It would be an interesting catch if that is the case.......

**Comment by therealkenc** [comment]
> it is only not working on wsl

On Real Linux:
```
$ mkdir nodetest
$ cd nodetest
$ dd if=/dev/zero of=my_fs bs=1024 count=30720
$ sudo losetup /dev/loop0 my_fs 
$ sudo mkfs.ntfs /dev/loop0
$ sudo mkdir -p /mnt/c
$ sudo mount /dev/loop0 /mnt/c
$ sudo chown ken:ken /mnt/c
$ mkdir -p /mnt/c/Users/David/Desktop/we/linux/node_modules
$ cd /mnt/c/User/David/Desktop/we
$ npm --version
5.5.1
$ npm init   # blah blah fill out form
$ npm install ./mnt/c/Users/David/Desktop/we/linux angular --save
$ npm ERR! code ENOLOCAL
$ npm ERR! Could not install from "mnt/c/Users/David/Desktop/we/linux" as it does not contain a package.json file.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ken/.npm/_logs/2017-12-17T02_17_35_697Z-debug.log
$ npm install /mnt/c/Users/David/Desktop/we/linux angular --save
npm ERR! code ENOLOCAL
npm ERR! Could not install from "linux" as it does not contain a package.json file.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ken/.npm/_logs/2017-12-17T02_19_32_112Z-debug.log
```

**Comment by davidmather** [comment]
Hey I got it working. It turned out that at some point in time npm broke that command. It is broken in a load of versions. Updating to the latest version of npm 5.6.0 using

`sudo npm install -g npm`

fixed the problem for me. Which I should have tried at the start and just totally forgot to. Now 

`npm install ./Linux`

Is working properly for me I also wrote a workaround for Linux just in case anybody runs into this command not working and absolutely needs to get it to work for a non-working npm version.

`npm install --prefix ./linux bootstrap --save && rsync --remove-source-files -avh --progress ./linux/node_modules ./ && rsync --remove-source-files -avh --progress ./linux/etc ./ && find ./linux -type d -empty -delete`

![capture](https://user-images.githubusercontent.com/13389155/34076501-2214437e-e2e1-11e7-8c8b-4edafc5839d7.PNG)

It is looking for a root package.json there but the command runs as expect.
